V2

Dieterbe opened the pull request.

Dieterbe committed:
the new stuff has arrived..

* implement a bunch of new formats
* rename things to make more sense
* godoc for all stuff

Dieterbe committed:
add test cases for time specifications

DanCech commented on README.md:
```diff
@@ -1 +1,51 @@
+* `YYYYMMDD`
+* `MM/DD/YY` 
+* `integer`: unix timestamp
+* `<monthname>+<num>` like `january+3` or `march+8`
```
`<monthname> <dayofmonth>` using `+` here is ambiguous since `+<duration>` is supported on any valid date specifier.

DanCech commented on datetime.go:
```diff
@@ -0,0 +1,233 @@
+			return uint32(now.Add(time.Duration(dur) * time.Second).Unix()), nil
+		}
+		return 0, err
+	}
```
graphite supports + and - syntax on any valid time specification, so this should probably be re-organized to take a similar approach by first trying to split out the modifier, then parsing the time specification, then applying the modifier.

Dieterbe committed:
don't conflate literal + with url-encoding +

+ can be literal + for like +2d and +7days
+ is also used as encoding for space, but that is a concern that is
outside the scope of this library. so we should use ' ' instead.

Dieterbe committed:
more ways to specify time

Dieterbe committed:
support shorter ways to specify a month

Dieterbe committed:
do it the dcech way

as a processing chain, step by step
so that you can list multiple modifiers in a row
+ more generic code

but it doesn't work yet because in cases like "timespec datespec",
many datespecs want to reset hours and minutes back to 0
(the code currently does this) but if it was preceeded by a timespec we
shouldn't.  I don't see a clean solution to this yet

Dieterbe commented:
@DanCech i tried your loopy way, see https://github.com/raintank/dur/tree/v2-dcech but see the commit msg i didn't see a nice way to handle `time date`. maybe i need to revisit this with a clear head, otherwise what do you think of the current code, good enough to go live?

DanCech committed:
make loop approach work

Dieterbe committed:
gofmt + fixes

Dieterbe commented:
so the dcech loopy way now works. it gets rid of 26 lines, at the expense of a bit of a slowdown, but this code should still be plenty fast in the bigger picture.
```
~/g/s/g/r/dur ❯❯❯ benchcmp v2.txt v2-loopy.txt
benchmark                                       old ns/op     new ns/op     delta
BenchmarkParseDateTime/now-8                    12.4          99.1          +699.19%
BenchmarkParseDateTime/-7d-8                    58.4          226           +286.99%
BenchmarkParseDateTime/now-7d10h3min-8          125           383           +206.40%
BenchmarkParseDateTime/today-8                  61.8          158           +155.66%
BenchmarkParseDateTime/monday-8                 955           564           -40.94%
BenchmarkParseDateTime/march_10-8               289           1029          +256.06%
BenchmarkParseDateTime/04:00_20110501-8         1106          767           -30.65%
BenchmarkParseDateTime/4pm_20110501-8           1027          705           -31.35%
BenchmarkParseDateTime/1234567890-8             44.3          141           +218.28%
BenchmarkParseDateTime/teatime_12/25/1998-8     1260          942           -25.24%
```

Dieterbe committed:
add benchmark

Dieterbe commented:
does this look good for merging and use in MT ?

DanCech reviewed (approved):
LGTM
